raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -386,6 +385,11 @@ private void onUpdateLeaderHighWatermark(
             // records still held in memory directly to the listener
             appendPurgatory.maybeComplete(highWatermark.offset(), currentTimeMs);
 
+            // After updating the high-watermark, complete all of the deferred
+            // fetch requests. This is always correct because all fetch request
+            // deferred have a HWM less or equal to the previous leader's HWM.
+            fetchPurgatory.completeAll(currentTimeMs);
```

Comment:
Want to make sure I understand why there's not a race here. If a new fetch arrives after we have updated the HWM (on L376), it's still okay to be completed because its guaranteed to not have the latest HWM. Is that right?